[mkdocs] docs: add namespace tutorials (#1810)
symbol/symbol: mkdocs/pages/en/devbook/namespaces/register-root-namespace.md

---
title: Register Root Namespace
---

# Registering a Root Namespace

> <Namespaces:> provide human-readable aliases for <mosaics:> and <accounts:>.

Reviewer: ```suggestion
<Namespaces:> provide human-readable aliases for <mosaics:> and <accounts:>,
which can be used instead of their cumbersome identifiers.
```

This tutorial shows how to register a root namespace and set its

> <lease duration:../../textbook/namespaces.md#duration>.

Reviewer: This is the first time "root namespace" is used so it needs to be defined.
The best way to do it is to just link to its glossary definition, but we don't have one.
So my suggestion is that you add definitions for "root namespace" and "subnamespace" in the Textbook page about Namespaces, right after the paragraph that starts with "The first part is called the root namespace."

Something like (please extend):

```markdown
Root Namespace
:   A <namespace:> that has no parent and can be used on its own.
    It can be used to group <subnamespaces:> together in a hierarchical manner.
Subnamespace
:   A <namespace:> that belongs to a parent <root namespace:>.
```

You can then remove the following paragraph in the Textbook and move its information (about the number of subnamespaces and their duration) to the above two definitions.

This would also fix the missing definition for <root namespace:> you have in the Registering a Subnamespace page.

> 

Reviewer: Add a last paragraph to the intro saying that once a namespace is registered, additional steps are required to link it to a mosaic or account, as explained in the Conclusions section.

- Create an <account:> to register the namespace, either
  [from code](../accounts/create-from-private-key.md) or
  [by using a wallet](../../userbook/wallet/create-account.md).
- Obtain <XYM:> to pay for the transaction and lease fees.
  See [Getting Testnet Funds from the Faucet](../accounts/testnet-faucet.md).

Additionally, review the [Transfer transaction](../transactions/transfer.md) tutorial to understand how
transactions are announced and confirmed.

## Full Code

{% import 'tutorial.jinja2' as tutorial with context %}

{{ tutorial.code_full('devbook/namespaces/register-root-namespace', ['py', 'js']) }}

## Code Explanation

### Setting Up the Account

{{ tutorial.code_snippet(['py:16:24', 'js:13:21']) }}

The snippet reads the signer's private key from the `SIGNER_PRIVATE_KEY` environment variable, which defaults to a
test key if not set.
The signer's address is derived from the public key.
This account will own the registered namespace.

### Fetching Network Time and Fees

{{ tutorial.code_snippet(['py:27:45', 'js:24:43']) }}

Network time and recommended fees are fetched from <get:/node/time> and <get:/network/fees/transaction> respectively,

> following the process described in the [Transfer Transaction](./transfer.md) tutorial.

Reviewer: ```suggestion
following the process described in the [Transfer Transaction](../transactions/transfer.md) tutorial.
```

### Building the Transaction

{{ tutorial.code_snippet(['py:47:59', 'js:45:57']) }}

The namespace registration transaction specifies:

* **Type:** Use `namespace_registration_transaction_v1`.

> * **Registration type:** The value `root` indicates a root namespace is being created.

Reviewer: You can mention the other option, and link to the other tutorial. For completeness.

* **Duration:** The number of blocks for which the namespace will be leased.
    The minimum duration is 86,400 blocks (approximately 30 days on mainnet), and the maximum is
    15,552,000 blocks (approximately 5 years).

* **Name:** The name of the root namespace.
    Names can only contain lowercase letters, numbers, hyphens, and underscores, must start with a letter or number,
    and can be at most 64 characters long.

>     To ensure the namespace name is unique across multiple runs, the example appends a timestamp to the name.

Reviewer: ```suggestion
    To ensure the namespace name is unique across multiple runs of the tutorial, a timestamp is added to the name.
    Applications will typically use a fixed name instead.
```

!!! note "Namespace lease fees"

    Registering a root namespace requires a lease fee in addition to the transaction fee.

>     The lease fee is proportional to the requested duration.

Reviewer: But I don't see the lease fee being paid in the transaction, only the standard transaction fee:
`transaction.fee = Amount(fee_mult * transaction.size)`

How come the code works?

The tutorial should show how to calculate the lease fee, but we have to figure out what is going on first.

Author: The network [deducts it automatically](https://github.com/symbol/symbol/blob/dev/client/catapult/plugins/txes/namespace/src/plugins/NamespaceRegistrationTransactionPlugin.cpp#L54) when the transaction is confirmed, so it is not needed to specify it in the transaction.

Author: Added some more details in the tutorial:

>     The network deducts it automatically when the transaction is confirmed, so you do not need to specify it in the transaction.

Reviewer: Ooooohhh... I understand. This was pretty confusing to me, because it's different from how transaction fees work.
I believe it will be confusing to other readers too, so it is worth highlighting the difference even more:

```md
    In addition to the standard [transaction fee](#fetching-network-time-and-fees),
    registering a namespace requires a [lease fee](../../textbook/namespaces.md#lease-fee) proportional to
    the requested duration.

    Unlike the transaction fee, the lease fee is **not** included in the transaction request.
    It is calculated and deducted automatically by the network from the **transaction signer’s account**
    when the registration transaction is confirmed.

    The amount of the lease fee can be calculated beforehand using the <get:/network/fees/rental> endpoint.
```

Make the same changes to the subnamespace tutorial.

    For details, see [Lease Fee](../../textbook/namespaces.md#lease-fee).

### Submitting the Transaction

{{ tutorial.code_snippet(['py:61:80', 'js:59:79']) }}

The transaction is signed and announced following the same process as in
[Creating a Transfer Transaction](../transactions/transfer.md#announcing-the-transaction).

{{ tutorial.code_snippet(['py:82:100', 'js:81:115']) }}

The code then waits for the transaction to be confirmed by polling the
<get:/transactionStatus/{hash}> endpoint until the status changes to `confirmed`.

### Retrieving the Namespace

{{ tutorial.code_snippet(['py:102:117', 'js:117:134']) }}

To verify the namespace was registered, the code retrieves it from the network
using the <get:/namespaces/{namespaceId}> endpoint and displays its properties.

The namespace ID is computed using <dy:IdGenerator.generateNamespaceId>.
This function applies a deterministic hashing algorithm to the namespace name,
producing the ID needed to query the namespace information.

The response includes:

> * **Registration type:** Shows `0` for a namespace.

Reviewer: ```suggestion
* **Registration type:** Shows `0` for a root namespace.
```

> * **Owner address:** The account that registered and owns the namespace.

Reviewer: There's only one account involved in this tutorial, however, I am getting 3 different numbers in the output:
```
Signer address: TCHBDENCLKEBILBPWP3JPB2XNY64OE7PYHHE32I
"signer_public_key": "3B6A27BCCEB6A42D62A3A8D02A6F0D73653215771DE243A63AC048A18B59DA29"
Owner address: 988E1191A25A88142C2FB3F69787576E3DC713EFC1CE4DE9
```
Why doesn't the last one (retrieved from the chain) match the first one?

Author: Fixed with:

> 		owner_address_hex = bytes.fromhex(namespace_info["ownerAddress"])
>		owner_address = Address(owner_address_hex)

The response returns the address hash without network byte nor checksum.

* **Start height:** The block height when the namespace was registered.

> * **End height:** The block height when the namespace will expire.

Reviewer: Why doesn't End height - Start height match 86400?

In my own experiments I get the exact same difference, 2880 blocks (1 day). Any idea where that comes from?

Author: It's due to the end grade period. For testnet, it is configured to one day.  See [namespaceGracePeriodDuration](https://reference.symboltest.net:3001/network/properties). 

Added some more context:

> * **Start and end heights** (lines 35-36): The namespace is active from block `2984442` to block `3073722`.
    The end height includes a [grace period](../../textbook/namespaces.md#duration) (1 day on <testnet:>,
    30 days on <mainnet:>) beyond the requested duration, giving owners time to renew before the namespace
    becomes available to others.

For root namespaces, you can calculate the remaining duration by subtracting the current block height from the end
height.

## Output

The output shown below corresponds to a typical run of the program.

```text
--8<-- 'devbook/namespaces/register-root-namespace.log'

> ```

Reviewer: Add line highlights and explain some of the interesting points :)
For example:
- The chosen namespace name (and how users can look it up in the explorer)
- The fee reflecting the transaction and the lease fees
- The id and the name (explain the difference, and that the name is hex-encoded)
- The star and end heights, and how the difference matches the requested lease (once we know why they don't match)

Author: Edited. Explained the interesting points in the conclusion,  but reduced the content in *Retrieve the namespace* to avoid repetition.

The transaction hash printed in the output can be used to search for the transaction
in the [Symbol Testnet Explorer](https://testnet.symbol.fyi/).

## Conclusion

This tutorial showed how to:

| Step                                                          | Related documentation                              |
| ------------------------------------------------------------- | -------------------------------------------------- |
| [Generate namespace ID](#building-the-transaction)            | <dy:IdGenerator.generateNamespaceId>               |

> | [Build a namespace registration](#building-the-transaction)   | <dy:SymbolTransactionFactory.create>               |

Reviewer: ```suggestion
| [Build a namespace registration transaction](#building-the-transaction)   | <dy:SymbolTransactionFactory.create>               |
```

| [Retrieve the namespace](#retrieving-the-namespace)           | <get:/namespaces/{namespaceId}>                    |

## Next Steps

Now that you have a root namespace, you can:

> - Link your namespace to a mosaic or account to create an alias

Reviewer: This point is important enough that I would promote it to an info box at the beginning of the Conclusions section.
Something like:

**⚠ The namespace is registered but it is not in use yet**
Namespaces are really useful when they can be used instead of the cumbersome identifiers required to refer to <mosaics:> and <accounts:>.
To be able to do this, the namespace needs to be linked to one such identifier, as explained in the guides linked below.

Author: Now added, but under "Retrieve the namespace".

This account will own the registered namespace.

### Fetching Network Time and Fees

{{ tutorial.code_snippet(['py:28:46', 'js:24:43']) }}

Network time and recommended fees are fetched from <get:/node/time> and <get:/network/fees/transaction> respectively,
following the process described in the [Transfer Transaction](../transactions/transfer.md) tutorial.

### Building the Transaction

{{ tutorial.code_snippet(['py:48:60', 'js:45:57']) }}

The namespace registration transaction specifies:

* **Type:** Namespace registration transactions use the type `namespace_registration_transaction_v1`.

* **Registration type:** The value `root` indicates a root namespace is being created.
    Use `child` to [register a subnamespace](./register-subnamespace.md) instead.

* **Duration:** The number of blocks for which the namespace will be leased.
    The minimum duration is 86,400 blocks (approximately 30 days on mainnet), and the maximum is
    15,552,000 blocks (approximately 5 years).

* **Name:** The name of the root namespace.
    Names can only contain lowercase letters, numbers, hyphens, and underscores, must start with a letter or number,
    and can be at most 64 characters long.

>     To ensure the namespace name is unique across multiple runs of the tutorial, a timestamp is added to the name.

Reviewer: ```suggestion
    To ensure the namespace name is unique across multiple runs of the tutorial, a timestamp is added to the name.
    In practice, programs would use a fixed name for their namespaces.
```

Let's be extra clear, as we were in the Metadata tutorial.
Add the same comment to the subnamespace tutorial.
